### PR TITLE
[ENH] Introducing BaseSegmenter for segmentation module

### LINKED ISSUE #802: [ENH] BaseSegmenter class
### Describe the feature or idea you want to propose

after the split of annotation module and the introduction of the experimental segmentation one, it would be good to settle on a standard interface for segmentation through the introduction of a BaseSegmenter. Currently, CLASP extends BaseSeriesAnnotator, GGS extends BaseEstimator and Hidalgo is in transformers.

### Describe your proposed solution

need to see if we can work with standard fit/predict, take conversions up to base class etc

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TonyBagnall**
#### Reference Issues/PRs
Fixes #802 #803
See #722 #699 #592 
    Segmenters take a single time series of size $m$ and return a segmentation.
    Series can be multivariate, with $d$ dimensions or channels.

    Design issues
    1. input and internal data format
    Univariate series:
        Numpy array, shape $(m,)$, $(m, 1)$ or $(1, m)$
            if self has no multivariate capability, all converted to 1D numpy $(m,)$
            if self has multivariate capability, converted to 2D numpy $(m,1)$ or $(1,
            m)$ depending on axis
        pandas DataFrame single column shape $(m,1)$, $(1,m)$ pandas Series shape $(m,)$
            if self has no multivariate capability, all converted to Series $(m,)$
            if self has multivariate capability, all converted to Pandas DataFrame
            shape $(m,1)$, $(1,m)$ depending on axis

    Multivariate series:
        Numpy array, shape $(m,d)$ or $(d,m)$
        pandas DataFrame $(m,d)$ or $(d,m)$

    2. Conversion and axis resolution for multivariate

    Conversion between numpy and pandas is handled by the base class. Sub classses
    can assume the data is in the correct format (determined by
    ``"X_inner_type"``, one of ``VALID_INNER_TYPES)`` and represented with the expected
    axis.

    Multivariate series are segmented along an axis determined by ``self.axis``. Axis
    plays two roles:
    1) the axis the segmenter expects the data to be in for its internal methods
    ``_fit`` and ``_predict``: 0 means each column is a time series channel $(m,d)$,
    1 means each row is a time series channel, sometimes called wide
    format, shape $(d,m)$. This should be set for a given child class through the
    BaseSegmenter constructor.

    2) The optional ``axis`` argument passed to the base class ``fit`` and ``predict``
    methods. If the data ``axis`` is different to the ``axis`` expected (i.e. value
    stored in ``self.axis``, then it is htransposed in this base class if self has
    multivariate capability.

    3. Segmentation representation

    Given a time series of 10 points with two change points found in position 4 and 8
    (lets index from 1 for clarity)

    The segmentation can be output in two forms:
    a) A list of change points: output example [4,8].
        This dense representation is the default behaviour, as it is the minimal
        representation.
    b) A list of integers of length n indicating the segment of each time point:
        output [0,0,0,1,1,1,1,2,2,2] or output [0,0,0,1,1,1,1,0,0,0]
        This sparse representation can be used to indicate shared segments (indicating
        segment 1 is somehow the same (perhaps in generative process) as segment 3.

    #TODO Add a tag to indicate which type of output
    Multivariate series are always segmented at the same points. If independent
    segmentation is required, fit a different segmenter to each channel.

### CODE REVIEW

**Comment by ermshaua** [comment]
Hi Tony.

Here are my thoughts on the BaseSegmenter.

1. Use fit/predict (current) or fit/transform

I think one can argue for one or the other. Personally, I prefer fit/predict, because I see segmentation as an annotation task, that "adds change points to the series" instead of "transforming it to another (collection) of series". This preference goes hand in hand with my thoughts on point 2. However, I also don't feel strongly.

2. Return array of same length (current) i.e. sparse format, change points (dense format) make it a class parameter (annotation model)

In the ClaSP implementation, "sparse" currently returns change points, while "dense" returns segment intervals. Both options make sense, personally I prefer just a numpy array of change points. There is a third option, however, which is a series of segment labels. GGS uses this representation. This cannot be returned by all segmentation algorithms, because ClaSP, for instance, only provides you the change points, not the segment labels (which may re-occur throughout the segmentation). 

So I think it would be a good choice to have a format parameter in the BaseSegmenter, where the user can choose one of the three options (if available). Default value should be the numpy array of change points. In the literature, sometimes methods define the first or last data point as a change point, I find this not to be necessary.

4. Applicable to series or to both series and collections?

Segmentation is typically defined as a series problem, not a collection of series task. Therefore, I would prefer the BaseSegmenter to be applicable to series. Currently, the user can provide different series in the fit and predict functions. To my understanding, this makes not too much sense. I'd rather prefer just a fit_predict function or have the series in the predict method optional and note that it is ignored.

Cheers, Arik

**Comment by TonyBagnall** [comment]
thanks Arik, so

File: aeon/segmentation/base.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+# allowed types for transformers - Series and Panel
+ALLOWED_INPUT_TYPES = [
+    "ndarray",
+    "Series",
+    "DataFrame",
+]
+
+
+class BaseSegmenter(BaseEstimator, ABC):
+    """Docstring.
```
Comment (baraline) [inline aeon/segmentation/base.py:55]: I guess the "Docstring."  should be replaced ?

File: aeon/segmentation/tests/test_base.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+testy = np.random.random(10)
+uni = []
+uni.append(testy)
+uni.append(pd.Series(testy))
+uni.append(pd.DataFrame(testy))
+# uni.append(np.random.random((10, 1)))
+testy = np.random.random((4, 10))
+multi = []
+multi.append(testy)
+multi.append(pd.DataFrame(testy))
```
Comment (baraline) [inline aeon/segmentation/tests/test_base.py:64]: Is this supposed to be here and not inside a test function ?

**Review by baraline — COMMENTED** [review]
Didn't find much to say on the segmentation logic as I only know the basics, apart from these minor issue the rest seems good to me !

File: aeon/segmentation/tests/test_base.py (hunk 2 above)
Comment (TonyBagnall) [reply aeon/segmentation/tests/test_base.py:64]: its used in two tests so made it global. Just doing a bit of test based development.

File: docs/api_reference/segmentation.rst [hunk 3]
```diff
@@ -13,6 +13,13 @@ contains algorithms and tools for time series segmentation.
...
+    BaseSegmenter
+    ClaSPSegmentater
+    InformationGainSegmenter
+    GreedyGaussianSegmenter
+    GreedyGaussianSegmenter
+    InformationGainSegmenter
+    DummySegmenter
+    EAggloSegmenter
+    HMMSegmenter
+    HidalgoSegmenter
```
Comment (MatthewMiddlehurst) [inline docs/api_reference/segmentation.rst:25]: There is some duplication here.

**Review by MatthewMiddlehurst — COMMENTED** [review]
I don't have much to comment on segmentation really, but had a quick look through and looks fine in principle. 

In the interest of moving things forward I would say @ anyone who may be interested to give comments on the changes and we can put it in after a bit if everyone is ok/no response.

It is all experimental, so can change easily.

File: docs/api_reference/segmentation.rst (hunk 3 above)
Comment (TonyBagnall) [reply docs/api_reference/segmentation.rst]: good spot!

File: docs/api_reference/annotation.rst [hunk 4]
```diff
@@ -7,15 +7,6 @@ The :mod:`aeon.annotation` module contains algorithms and tools
 for time series annotation. This no longer includes anomaly/outlier
 detection and time series segmentation, which are now in their own module.
```
Comment (baraline) [inline docs/api_reference/annotation.rst:8]: Minor point, but a `:mod:` link to the new module would be nice

**Review by baraline — DISMISSED** [review]
Couldn't find much to say on this PR apart from a very minor comment in the docs, LGTM !

File: aeon/segmentation/base.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    1) the axis the segmenter expects the data to be in for its internal methods
+    ``_fit`` and ``_predict``: 0 means each column is a time series channel `(m,d)`,
+    1 means each row is a time series channel, sometimes called wide
+    format, shape `(d,m)`. This should be set for a given child class through the
+    BaseSegmenter constructor.
+
+
+    2) The optional ``axis`` argument passed to the base class ``fit`` and ``predict``
+    methods. If the data ``axis`` is different to the ``axis`` expected (i.e. value
+    stored in ``self.axis``, then it is htransposed in this base class if self has
```
Comment (lmmentel) [inline aeon/segmentation/base.py:67]:
```suggestion
    stored in ``self.axis``, then it is transposed in this base class if self has
```

File: aeon/segmentation/base.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    segmentation is required, fit a different segmenter to each channel.
+
+    Parameters
+    ----------
+    n_segments : int, default = 2
+        Number of segments to split the time series into. If None, then the number of
+        segments needs to be found in fit.
+    axis : int, default = 0
+        Axis along which to segment if passed a multivariate series (2D input). If axis
+        is 0, it is assumed each row is a time series and each column is a channel.
```
Comment (lmmentel) [inline aeon/segmentation/base.py:95]: Not sure what the difference here is between "time series" and "channel", maybe "time point" or "time instance" would be better?

**Review by lmmentel — COMMENTED** [review]
Looks good :raised_hands:  Very nice that all segmentation algos now live in one place :)

File: aeon/segmentation/_dummy.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        ----------
+        X : 2D np.ndarray
+            Time series of shape `(n_channels, series_length)`
+        y : np.ndarray or None, default = None
+
+        Returns
+        -------
+        self : reference to self.
+        """
+        self.n_segments_ = self.n_segments
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/segmentation/_dummy.py:40]: think I got a bit mixed up here with number of segments in the base class. I say it defaults to 2 but then default to None. I think default to 2 is reasonable. I will rework it.

File: aeon/segmentation/base.py (hunk 6 above)
Comment (TonyBagnall) [reply aeon/segmentation/base.py:95]: agreed

File: aeon/segmentation/_dummy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/segmentation/_dummy.py]: reworked the n_segments to default to 2

**Comment by TonyBagnall** [comment]
hopefully addressed comments, thanks for the input. I'd appreciate it if someone could read the docstrings, the whole axis thing does my head in, and it was incorrect before

**Review by lmmentel — APPROVED** [review]

**Pull request merged**